### ISSUE TITLE
Support taint propagation from MachinePool to the underlying nodes

### ISSUE THREAD

**Issue by Archisman-Mridha** [kubernetes-sigs/cluster-api#11175]
### What would you like to be added (User Story)?

It'll be very helpful if we have this feature support :
> Declare what taints should be applied to the underlying node(s) in the Machine / MachineDeployment / MachinePool spec.

### Detailed Description

Especially, when I'm creating a MachinePool. I want to specify taints in the MachinePool spec, and those taints will be applied to each node managed by that MachinePool.
Those nodes will then be reserved for specific types of workloads which have tolerations against those node taints.

**This is will help manage node taints in a declarative manner using GitOps.**

### Anything else you would like to add?

I've come up with a Plan Of Action here : https://github.com/Archisman-Mridha/cluster-api/commit/a3a807b05ba5808b8c45ca131e1ee6f99cbdce55

### Label(s) to be applied

/kind feature
One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels.

**Comment by fabriziopandini** [comment]
If it is only for machine creation, the current solution is to rely on the capabilities of bootstrap providers, but I'm aware of the fact that not all the MP implementations use a bootstrap providers.

If we are looking for a more generic story, that do not apply only on creation, this requires some design, especially for figuring out the interactions between the new "taint propagation" feature and taints defined by bootstrap providers

In order to define next steps, might be worth taking a look at https://cluster-api.sigs.k8s.io/developer/architecture/controllers/metadata-propagation and underlying proposal: 
- https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20220927-label-sync-between-machine-and-nodes.md
- [20221003-In-place-propagation-of-Kubernetes-objects-only-changes.md](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20221003-In-place-propagation-of-Kubernetes-objects-only-changes.md)

cc @jackfrancis @mboersma @willie-yao 

@elmiko this discussion might be useful to revive https://github.com/kubernetes-sigs/cluster-api/issues/7685

**Comment by Archisman-Mridha** [comment]
@fabriziopandini Thanks for the resources you linked. I wasn't aware of the `JoinConfiguration.NodeRegistration.Taints` field in the `KubeadmConfigTemplate` custom resource. I can use that to apply taint to newly minted worker nodes (managed by a MachinePool in my case). Currently this is what my company needs.

However there will be a downside - we can't **in-place change** (add / remove) taints for those worker nodes. And I'm interested in contributing the solution for this. We want the user to be able to **specify the taints in the Machine spec**. ClusterAPI will then sync those taints between the Machine resource and the corresponding Kubernetes node.

That's what I'm trying to implement here : https://github.com/Obmondo/cluster-api/commit/ab19c36069d220318e2e2caa7159a2a700ad9ec8. Can you please give any opinion on whether I'm going in the right direction or not ?

**Comment by JoelSpeed** [comment]
> However there will be a downside - we can't in-place change (add / remove) taints for those worker nodes

Removing taints becomes problematic if you have the potential for multiple actors to add taints (which is always the case, think about cordoning a node using kubectl).

Typically in cases where we've had fields like this before, you'd omit the removal side of things, and the taints would be added if they are in the spec, and manually removed if removed from spec.

However, now that we have SSA, in theory we can make this work. If we were to use SSA to apply the taints, then whenever we apply the list of taints from our side, any that _we applied_ (and therefore own according to field managers) that we no longer specify would be removed for us. If they are later re-added by something other, then they would be the field owner now and we wouldn't know that we previously owned it, so, this _could_ be safe if we use SSA

**Comment by nrb** [comment]
/remove-lifecycle stale
/assign

I'm looking into this and #7409. For annotations, most of the propagation amongst Kubernetes objects was handled via https://github.com/kubernetes-sigs/cluster-api/issues/7731, but such plumbing's not yet been done for taints.

**Comment by nrb** [comment]
I think this might be useful outside of machinepools, too, but I agree with Fabrizio that it should be specified throughout the types, not just a subset.

I'm taking a look at the history of taints in CAPI; I know that `machine.Spec.Taints` existed in v1alpha1, but was moved into bootstrap providers with v1alpha2. https://github.com/kubernetes-sigs/cluster-api/pull/997 doesn't explicitly say why that is, but I suspect it's because applying taints was considered a one-time operation.

**Comment by sbueringer** [comment]
I assume taints were moved to the bootstrap provider so the taints are set when creating/registering the Node. If taints are added later Pods that don't tolerate the taint might have already been scheduled to the Node

**Comment by JoelSpeed** [comment]
> If taints are added later Pods that don't tolerate the taint might have already been scheduled to the Node

To be fair, that is the purpose of the NoExecute style taints, where pods can then be evicted from a Node because they no longer match the taints on it.

> taints are atomic + how to deal with multiple actors adding/removing taints (which is always the case, think about cordoning a node using kubectl).

Are we talking about continuous reconciliation of taints, or just adding them when we first see the node object created? I think we need to be clear on which of these we want to support before we get too far

**Comment by sbueringer** [comment]
> To be fair, that is the purpose of the NoExecute style taints, where pods can then be evicted from a Node because they no longer match the taints on it.

Makes sense. I guess the only limitation then is that if you add the taints later you cannot ensure that the Pods never *ran* on these Nodes

Btw, probably this can be mitigated, we already have an optional contract for bootstrap providers to add an additional `node.cluster.x-k8s.io/uninitialized:NoSchedule` taint. This taint is currently removed after we sync labels/annotations for the first time.

> continuous reconciliation of taints

I think if we do it we should continuously reconcile them, like labels and annotations. I think if we treat taints differently it would be very surprising to users.

**Comment by JoelSpeed** [comment]
> Makes sense. I guess the only limitation then is that if you add the taints later you cannot ensure that the Pods never ran on these Nodes

So this _could_ be an issue for us if we are saying that the Node comes up, and pods are scheduled, and then we taint the node. There is a short (?) period where a pod may schedule and we haven't sorted the taints yet, hmm 🤔 

> Btw, probably this can be mitigated, we already have an optional contract for bootstrap providers to add an additional node.cluster.x-k8s.io/uninitialized:NoSchedule taint. This taint is currently removed after we sync labels/annotations for the first time.

Something similar to this would make a lot of sense. I guess the bootstrap provider uses the `kubelet` `--register-with-taints` to achieve this? We would have to make sure the bootstrap providers also add a taint for the taint propagation 

> I think if we do it we should continuously reconcile them, like labels and annotations. I think if we treat taints differently it would be very surprising to users.

Ack, then we should think very carefully about how we make updates to taints atomically and don't fight with other actors. In theory if we use the RV (if using SSA/patch) as precondition to the request that _should_ prevent us from updating the list if someone else has written to the object since. This may need some testing though as I expect Nodes get written to frequently and so this could cause conflict/retry loops moreso than another object

**Comment by sbueringer** [comment]
> Something similar to this would make a lot of sense. I guess the bootstrap provider uses the kubelet --register-with-taints to achieve this? We would have to make sure the bootstrap providers also add a taint for the taint propagation

IIRC. Yes. It's done by kubeadm in the case of CABPK

**Comment by sbueringer** [comment]
> I think if we do it we should continuously reconcile them, like labels and annotations. I think if we treat taints differently it would be very surprising to users.

Hm. I think I"ll take that partially back. We have to think about the trade-offs.

1. If this is done with in-place update it's nice that we can change taints on existing Nodes without having to recreate the Nodes, but this also means that the taints only apply from this point forward

2. If we do it with replacement we can guarantee that on the new Nodes the Pods were never ran.

Not sure what is more important for taint use cases.

CABPK with MDs basically implements case 2 today

**Comment by JoelSpeed** [comment]
I think given that users of Kubernetes without Cluster API would expect taints on Nodes to be mutable without having to replace the Nodes, if a user modified the taints on a specific Machine, I think it would be acceptable to have those mutated in place on the existing Node.

If they mutated the Deployment spec however, I think then that would be a signal to roll out new nodes, right?

**Comment by sbueringer** [comment]
> If they mutated the Deployment spec however, I think then that would be a signal to roll out new nodes, right?

Depends. Today we have fields that trigger rolllouts (Kubernetes version) and fields that are mutated in place (labels, annotations, drain timeout, etc.)

**Comment by nrb** [comment]
> fields that are mutated in place (labels, annotations, drain timeout, etc.)

The main difference between the labels/annotations and taints, in my mind, is that adding taints later will (probably) trigger evictions. I'm not aware of anything in stock Kubernetes that would do so with labels/annotations, but of course someone could write their own controllers for that behavior.

**Comment by nrb** [comment]
I've uploaded the proposal here: https://github.com/kubernetes-sigs/cluster-api/pull/12329. Let's continue the discussion there

**Comment by ujjwal405** [comment]
@nrb  is this issue still persist??

**Comment by sbueringer** [comment]
I think we can and should close this issue in favor of https://github.com/kubernetes-sigs/cluster-api/issues/12908

**Comment by sbueringer** [comment]
@fabriziopandini @nrb what do you think?
